url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/CameraAgent.java

```diff
@@ -0,0 +1,554 @@
+	};
+
+	public enum KitStatus {
+		NOT_READY, DONE, PICTURE_BEING_TAKEN, FAILED
+	};
+
+	public class MyKit {
+		public Kit kit;
+		public KitStatus ks;
+		public boolean kitDone;
+
+		public MyKit(Kit k) {
+			kit = k;
+			ks = KitStatus.NOT_READY;
+			kitDone = false;
+		}
+	}
+
+	public class MyNest {
+		public NestAgent nest;
+		public NestStatus state;
+		public int numFilledSnapshot = 0;
+
+		public MyNest(NestAgent nest) {
+			this.nest = nest;
+			this.state = NestStatus.NOT_READY;
+		}
+	}
+
+	public CameraAgent(String name) {
+		super();
+		this.name = name;
+
+		timer = new Timer();
+		timer.scheduleAtFixedRate(new TimerTask() {
+			// Fires every 3.001 seconds.
+			@Override
+			public void run() {
+				print("Waking up");
+				stateChanged();
+			}
+		}, System.currentTimeMillis(), 3000);
+	}
+
+	/*
+	 * Messages
+	 */
+	@Override
+	public void msgInspectKit(Kit kit) {
+		print("Received msgInspectKit");
+		mk = new MyKit(kit);
+		stateChanged();
+	}
+
+	// Should no longer be called in v2. Change to timer.
+	@Override
+	public void msgIAmFull(Nest nest) {
+		// synchronized (nests) {
+
+		// print("Received msgIAmFull from nest with type " + ((NestAgent)
+		// nest).currentPartType);
+		for (MyNest n : nests) {
+			if (n.nest == nest) {
+				if (n.state == NestStatus.NOT_READY || n.state == NestStatus.STARTED_TIMER) {
+					n.state = NestStatus.READY;
+				} else if (n.state == NestStatus.WAITING_TO_RE_PHOTOGRAPH) {
+					n.state = NestStatus.READY_TO_RE_PHOTOGRAPH;
+				} else if (n.state == NestStatus.WAITING_TO_RE_PHOTOGRAPH_AGAIN) {
+					n.state = NestStatus.READY_TO_RE_PHOTOGRAPH_AGAIN;
+				}
+				break;
+			}
+		}
+		// }
+		stateChanged();
+	}
+
+	@Override
+	public void msgResetSelf() {
+		print("Reseting Self");
+		// synchronized (nests) {
+		for (MyNest n : nests) {
+			n.state = NestStatus.NOT_READY;
+		}
+		// }
+		mk = null;
+		timer.cancel();
+		timer = new Timer();
+		timer.scheduleAtFixedRate(new TimerTask() {
+			// Fires every 3.001 seconds.
+			@Override
+			public void run() {
+				print("Waking up");
+				stateChanged();
+			}
+		}, System.currentTimeMillis(), 3000);
+		// stateChanged();
+	}
+
+	@Override
+	public void msgTakePictureNestDone(NestGraphics nest, boolean done, NestGraphics nest2, boolean done2) {
+		print("Received msgTakePictureNestDone from graphics");
+		boolean found1 = false;
+		boolean found2 = false;
+		// synchronized (nests) {
+		for (MyNest n : nests) {
+			if (n.nest.nestGraphics == nest) {
+				// In v0 all parts are good parts
+				print("nest " + nests.indexOf(n) + " photographed");
+				n.numFilledSnapshot = n.nest.currentParts.size();
+
+				if (n.state == NestStatus.PHOTOGRAPHING) {
+					n.state = NestStatus.PHOTOGRAPHED;
+				} else if (n.state == NestStatus.RE_PHOTOGRAPHING_ONCE) {
+					n.state = NestStatus.RE_PHOTOGRAPHED_ONCE;
+				} else if (n.state == NestStatus.RE_PHOTOGRAPHING_TWICE) {
+					n.state = NestStatus.RE_PHOTOGRAPHED_TWICE;
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 154:160, 171:177, 470:476, 477:483.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.